File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root: return False
        
        def isSame(root, subRoot):
            if not root and not subRoot: return True
            elif root and subRoot:
                return root.val == subRoot.val and isSame(root.left, subRoot.left) and isSame(root.right, subRoot.right)
            
        if isSame(root,subRoot): return True
        
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
                
        

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function(root, subRoot) {
    /*
    dfs
    */
   
    if(!root) return false;
    if(areSame(root, subRoot) || !subRoot) return true;
    return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    
    function areSame(t1,t2){
        if(!t1 && !t2) return true;
        if(!t1 || !t2) return false;
        if(t1.val === t2.val) {
            return areSame(t1.left, t2.left) && areSame(t1.right, t2.right)
        }
    }
};